### ISSUE TITLE
stale events are pushed repeatedly

### ISSUE THREAD

**Issue by Huang-Wei** [kubernetes/client-go#338]
I'm using a tweaked `examples/workqueue/main.go` (v5.0.1) to reproduce a weird issue: stale events are pushed to client repeatedly even when an API object has been deleted:

**Env:**

- k8s server: minikube v1.7.4
- client: `examples/workqueue/main.go` - replace `Pod` with `Service`

**Reproduce steps:**

1. run the main.go - `./main --kubeconfig=/Users/wei.huang1/.kube/config --v=4 --logtostderr=true`
1. create a shell script to do (use kubectl) following things **in parallel**:
    1. create 3 different `Namespace`s
    1. create 3 nginx `Service`s along with nginx `Deployment`s in each Namespace 
    1. delete all 3 namespaces **immediately**

**Log:**

```
I1201 14:58:09.938273   46973 main.go:121] Starting Service controller
I1201 14:58:09.938536   46973 reflector.go:202] Starting reflector *v1.Service (0s) from k8s.io/client-go/examples/workqueue/main.go:123
I1201 14:58:09.938552   46973 reflector.go:240] Listing and watching *v1.Service from k8s.io/client-go/examples/workqueue/main.go:123
I1201 14:58:10.041648   46973 shared_informer.go:116] caches populated
Sync/Add/Update for Service kubernetes
Sync/Add/Update for Service kube-dns
Sync/Add/Update for Service kubernetes-dashboard
Sync/Add/Update for Service alertmanager
Sync/Add/Update for Service prometheus
Sync/Add/Update for Service nginx-svc
Sync/Add/Update for Service nginx-svc
Sync/Add/Update for Service nginx-svc
Service 590-2-10871/nginx-svc does not exist anymore
Service 590-1-12560/nginx-svc does not exist anymore
Service 590-3-7850/nginx-svc does not exist anymore
# logs until here is good
---
# interesting log begins
I1201 15:06:10.957585   46973 reflector.go:421] k8s.io/client-go/examples/workqueue/main.go:123: Watch close - *v1.Service total 6 items received
Sync/Add/Update for Service nginx-svc
Service 590-1-12560/nginx-svc does not exist anymore
I1201 15:15:52.960785   46973 reflector.go:421] k8s.io/client-go/examples/workqueue/main.go:123: Watch close - *v1.Service total 4 items received
Sync/Add/Update for Service nginx-svc
Service 590-1-12560/nginx-svc does not exist anymore
I1201 15:24:11.964377   46973 reflector.go:421] k8s.io/client-go/examples/workqueue/main.go:123: Watch close - *v1.Service total 4 items received
Sync/Add/Update for Service nginx-svc
Service 590-1-12560/nginx-svc does not exist anymore
I1201 15:31:22.967810   46973 reflector.go:421] k8s.io/client-go/examples/workqueue/main.go:123: Watch close - *v1.Service total 4 items received
Sync/Add/Update for Service nginx-svc
Service 590-1-12560/nginx-svc does not exist anymore
I1201 15:38:29.970636   46973 reflector.go:421] k8s.io/client-go/examples/workqueue/main.go:123: Watch close - *v1.Service total 4 items received
Sync/Add/Update for Service nginx-svc
Service 590-1-12560/nginx-svc does not exist anymore
I1201 15:46:55.972923   46973 reflector.go:421] k8s.io/client-go/examples/workqueue/main.go:123: Watch close - *v1.Service total 4 items received
Sync/Add/Update for Service nginx-svc
Service 590-1-12560/nginx-svc does not exist anymore
I1201 15:52:14.976354   46973 reflector.go:421] k8s.io/client-go/examples/workqueue/main.go:123: Watch close - *v1.Service total 4 items received
Sync/Add/Update for Service nginx-svc
Service 590-1-12560/nginx-svc does not exist anymore
```

So my question is: is it a bug that the Add/Delete events (590-1-12560/nginx-svc) are triggered over and over even if they have been deleted? and might be something wrong in namespace garbage collector?

**Comment by liggitt** [comment]
Update events during namespace deletion is not necessarily a bug. Objects that still exist can be updated during namespace deletion. Are you getting add or update events for the services?

**Comment by Huang-Wei** [comment]
@liggitt yes, updating events is not a bug - esp. in the case of client-go resync every ~10 minutes.

But the log above shows that it's **Add and Delete** events of a particular service (nginx-svc Service in namespace 590-1-12560):

```
I1201 15:06:10.957585   46973 reflector.go:421] k8s.io/client-go/examples/workqueue/main.go:123: Watch close - *v1.Service total 6 items received
Sync/Add/Update for Service nginx-svc <== Add event
Service 590-1-12560/nginx-svc does not exist anymore <== Delete event
```

Any hints on this?

**Comment by Huang-Wei** [comment]
reproduced again. this time I tried the test slightly differently:

1. create a shell script to do (use kubectl) following things in parallel:
    - create 3 different Namespaces
    - create 3 nginx Services along with nginx Deployments in each Namespace
1. manually delete 3 nginx Services one by one - `kubectl delete svc1 -n ns1 && kubectl delete svc2 -n ns2 && kubectl delete svc3 -n ns3`

So it looks like irrelevant with namespace garbage collector.

**Comment by liggitt** [comment]
same root cause as #58545, fixed in #58547

delete events were sending the incorrect resource version. that would confuse the informer, which keeps track of the last resourceVersion it sees, so it can restart watches from that resourceVersion.

**Comment by Huang-Wei** [comment]
thanks @liggitt ! It seems the fix goes to apiserver side, so which Kubernetes release/tag contains this fix?

**Comment by liggitt** [comment]
It has been picked to release branches and will be available in 1.7.13, 1.8.8, and 1.9.3
